### DISCUSSION TITLE
Dns requests not working

### DISCUSSION THREAD

**Discussion by matinkg** [v2fly/v2ray-core#2444] [Q&A]
Hello, I am looking to use v2ray-core as a bridge. Currently, I have a socks5 proxy set up on my localhost and I want v2ray to proxy all requests to that proxy. In essence, I want the following setup: client => vmess => socks5 => destination.

I have created a configuration that achieves this, but unfortunately, DNS requests are not functioning properly. As a result, I can only connect to servers using their IP addresses. I was hoping someone could provide some guidance on what steps I should take to resolve this issue. Thank you.

Here is my server v2ray configuration:

```
{
    "stats": {},
    "api": {
        "tag": "api",
        "services": [
            "StatsService"
        ]
    },
    "log": {
        "loglevel": "warning"
    },
    "policy": {
        "levels": {
            "0": {
                "statsUserUplink": true,
                "statsUserDownlink": true
            }
        },
        "system": {
            "statsInboundUplink": true,
            "statsInboundDownlink": true,
            "statsOutboundUplink": true,
            "statsOutboundDownlink": true
        }
    },
    "inbounds": [
        {
            "tag": "in-vmess",
            "port": 5555,
            "protocol": "vmess",
            "settings": {
                "clients": [
                    {
                        "id": "xxxx",
                        "level": 0,
                        "email": "xxxx"
                    }
                ]
            },
            "streamSettings": {
                "network": "ws",
                "wsSettings": {
                    "path": "/api"
                }
            }
        },
        {
            "listen": "127.0.0.1",
            "port": 10085,
            "protocol": "dokodemo-door",
            "settings": {
                "address": "127.0.0.1"
            },
            "tag": "api"
        }
    ],
    "outbounds": [
        {
            "protocol": "freedom",
            "tag": "dflt"
        },
        {
            "protocol": "socks",
            "settings": {
                "servers": [
                    {
                        "address": "127.0.0.1",
                        "port": 8889
                    }
                ]
            },
            "tag": "out-socks"
        }
    ],
    "dns": {
        "servers": [
            "https+local://1.1.1.1/dns-query",
            "8.8.8.8",
            "8.8.4.4",
            "1.1.1.1",
            "1.0.0.1",
            "4.2.2.4",
            "localhost"
        ]
    },
    "routing": {
        "rules": [
            {
                "inboundTag": [
                    "api"
                ],
                "outboundTag": "api",
                "type": "field"
            },
            {
                "type": "field",
                "inboundTag": [
                    "in-vmess"
                ],
                "outboundTag": "out-socks"
            }
            
        ],
        "domainStrategy": "AsIs"
    }
}
```

**Comment by mydogshitgold** [comment]
Do you mean that your DNS requests does not work at all?

The built-in DNS module of v2ray functions as an inbound, necessitating the allocation of a tag and the routing of its traffic unto an appropriate outbound; failure to do so would result in the diversion of said traffic into the default outbound, namely “dflt” in your case. 
Inability to resolve domain names may result from the lack of direct access to an untainted DNS server by your server.

---

Or, do you mean that the domain name of your v2ray server is usable in the configuration of the v2ray client?

V2ray does not conveniently proxy all DNS requests originating from your OS, unless you explicitly setup a DNS inbound and adjust your OS to utilize it. Additionally, proxying DNS request regarding your v2ray server’s domain name are discouraged, as the inevitable “chicken or egg” conundrum may arise; rather, you should locally resolve the domain names of your servers.

---

If my interpretation proves to be inaccurate, can you give more context?

**Reply by matinkg** [reply]
I am using two clients. Nekoray for windows and V2rayNG for android.
Using Nekoray on windows, DNS seems working but the problem is it is not proxying to the socks server thus I get the wrong IP for banned websites.
Here is an example of v2ray server log when I try to open a banned website:
```
2023/03/30 22:38:30 5.119.88.11:64552 accepted tcp:10.10.34.35:443 [out-socks]
```
10.10.34.35 is the IP of the block page for banned websites

Using V2rayNG on android, the DNS is not working at all. When I open a URL it gives "DNS address could not be found" error.
Here is the example of v2ray server log when I try to open a URL:
```
2023/03/30 22:41:18 5.119.88.11:2056 accepted udp:1.1.1.1:53 [out-socks]
2023/03/30 22:41:18 [Warning] [1782737717] app/proxyman/outbound: failed to process outbound traffic > proxy/socks: failed to establish connection to server > EOF
```

On both android and windows, I can use telegram, Instagram, and other applications without any problem. The problem happens when I want to open a URL in the browser and I'm mostly sure it is because of the DNS.

I guess I should route all dns requests to the socks proxy but I don't know how

**Reply by mydogshitgold** [reply]
The IP address `10.10.34.35` is considered a bogon address that has been frequently utilized in Iran to intentionally corrupt DNS cache. If the domestic DNS servers are naughty, you may want to establish a DNS inbound and configure your operating system to utilize it. 

> I guess I should route all dns requests to the socks proxy but I don't know how

Since most socks proxies does not support UDP, you should either use DNS over TCP or DoH.

I think the this [article](https://guide.v2fly.org/basics/dns.html) has the potential to provide answers to your queries, despite being written in Mandarin, as the translation appears to be of good quality.

[DNS in V2Ray (translation).pdf](https://github.com/v2fly/v2ray-core/files/11117356/DNS.in.V2Ray.pdf)

---
آرزوی سعادت برای شما دارم و امیدوارم که آزادی در ایران برتری یابد.

**Reply by matinkg** [reply]
Thanks for your reply
I read the article you provided but unfortunately I get nothing out of it. I don't know what exactly the problem is and v2ray documentation is a little hard for me to understand.
Is it possible for you to explain more or provide a config that fix my issue?

Thanks

**Comment by lhbdhr** [comment]
In fact, you don't need to set dns in the server configuration file.
dns is used to determine whether the website accessed by your client is through a proxy, and only needs to be configured on the client.
```json
{
  "log": {
    "loglevel": "error",
    "access": "/etc/v2fly/logs/access.log",
    "error": "/etc/v2fly/logs/error.log"
  },
  "inbounds": [
    {
      "port": 5555,
      "protocol": "vmess",
      "settings": {
        "clients": [
          {
            "id": "xxxx",
            "level": 0,
            "email": "xxxx"
          }
        ]
      },
      "streamSettings": {
        "network": "ws",
        "wsSettings": {
          "path": "/api"
        }
      }
    }
  ],
  "outbounds": [
    { "protocol": "freedom", "tag": "freedom" },
    {
      "protocol": "socks",
      "tag": "socks-out",
      "settings": {
        "servers": [
          {
            "address": "socks-out",
            "port": 9091
          }
        ]
      }
    }
  ],
  "routing": {
    "rules": [
      {
        "type": "field",
        "domain": ["domain:ai.com", "domain:openai.com", "domain:ip.gs"],
        "outboundTag": "socks-out"
      },
      {
        "type": "field",
        "network": "tcp,udp",
        "outboundTag": "freedom"
      }
    ]
  }
}
```
My server does not allow access to openai, I use cloudflare's warp,
Warp can open the socks service, so I just used the socks proxy,
The most important part is routing. This configuration file is set to access through socks if openai is accessed.
Routing is matched according to the order. The last rule is that if it is tcp or udp traffic (lol, it contains all traffic), it is accessed through freedom, which is the server itself.
That is to say, all traffic that does not match the previous routing rules will be accessed through the server itself.

**Reply by iamyegor** [reply]
Did you find the solution?

**Comment by oops10101** [comment]
Hey there I'm currently struggling with the exact same thing did you find any solution? thank you in advance

**Comment by oops10101** [comment]
> Hello, I am looking to use v2ray-core as a bridge. Currently, I have a socks5 proxy set up on my localhost and I want v2ray to proxy all requests to that proxy. In essence, I want the following setup: client => vmess => socks5 => destination.
> 
> I have created a configuration that achieves this, but unfortunately, DNS requests are not functioning properly. As a result, I can only connect to servers using their IP addresses. I was hoping someone could provide some guidance on what steps I should take to resolve this issue. Thank you.
> 
> Here is my server v2ray configuration:
> 
> ```
> {
>     "stats": {},
>     "api": {
>         "tag": "api",
>         "services": [
>             "StatsService"
>         ]
>     },
>     "log": {
>         "loglevel": "warning"
>     },
>     "policy": {
>         "levels": {
>             "0": {
>                 "statsUserUplink": true,
>                 "statsUserDownlink": true
>             }
>         },
>         "system": {
>             "statsInboundUplink": true,
>             "statsInboundDownlink": true,
>             "statsOutboundUplink": true,
>             "statsOutboundDownlink": true
>         }
>     },
>     "inbounds": [
>         {
>             "tag": "in-vmess",
>             "port": 5555,
>             "protocol": "vmess",
>             "settings": {
>                 "clients": [
>                     {
>                         "id": "xxxx",
>                         "level": 0,
>                         "email": "xxxx"
>                     }
>                 ]
>             },
>             "streamSettings": {
>                 "network": "ws",
>                 "wsSettings": {
>                     "path": "/api"
>                 }
>             }
>         },
>         {
>             "listen": "127.0.0.1",
>             "port": 10085,
>             "protocol": "dokodemo-door",
>             "settings": {
>                 "address": "127.0.0.1"
>             },
>             "tag": "api"
>         }
>     ],
>     "outbounds": [
>         {
>             "protocol": "freedom",
>             "tag": "dflt"
>         },
>         {
>             "protocol": "socks",
>             "settings": {
>                 "servers": [
>                     {
>                         "address": "127.0.0.1",
>                         "port": 8889
>                     }
>                 ]
>             },
>             "tag": "out-socks"
>         }
>     ],
>     "dns": {
>         "servers": [
>             "https+local://1.1.1.1/dns-query",
>             "8.8.8.8",
>             "8.8.4.4",
>             "1.1.1.1",
>             "1.0.0.1",
>             "4.2.2.4",
>             "localhost"
>         ]
>     },
>     "routing": {
>         "rules": [
>             {
>                 "inboundTag": [
>                     "api"
>                 ],
>                 "outboundTag": "api",
>                 "type": "field"
>             },
>             {
>                 "type": "field",
>                 "inboundTag": [
>                     "in-vmess"
>                 ],
>                 "outboundTag": "out-socks"
>             }
>             
>         ],
>         "domainStrategy": "AsIs"
>     }
> }
> ```

Hey there I'm currently struggling with the exact same thing did you find any solution? thank you in advance

**Comment by jacksonlie1990** [comment]
I'm currently struggling with the exact same problem!
